### PR TITLE
Allow dictionary inputs for wps_parameters

### LINKED ISSUE #23: Allow wps_parameters to use dictionary input
Using `wps_convolution.py` as a template, update `wps_parameters.py` to use a dictionary instead of configuration file for inputs.

### PR DESCRIPTION
**Pull request by sum1lim**
This PR resolves #23 

Currently, `RVIC parameters` module only allows inputs through pre-generated `.cfg` file. It degrades usability of the software. Therefore, dictionary inputs are allowed by `osprey/wps_parameters` and processed into `.cfg` files. Default values are set for many of the parameters (keys) referring to [Lee's note](https://pcic.uvic.ca/confluence/pages/viewpage.action?spaceKey=CSG&title=Meeting+Notes+2018-05-09+-+RVIC+demo) in confluence.

Example dictionary inputs are added to both `jupyter lab` demo and `test_wps_parameters`

### CODE REVIEW

File: osprey/processes/wps_convolution.py [hunk 1]
```diff
@@ -127,17 +127,16 @@ def _handler(self, request, response):
...
-                convolution,
-                version,
-                config,
-                config_file_builder(self.workdir, config, self.config_template),
+            config = config_hander(
+                self.workdir, convolution.__name__, unprocessed, self.config_template
             )
 
+        run_rvic(
+            self.workdir, convolution, version, config,
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/processes/wps_convolution.py:137]: Now we don't. But it shouldn't be a problem for now, and I am working on this another [PR](https://github.com/pacificclimate/osprey/pull/32).

File: osprey/processes/wps_convolution.py [hunk 2]
```diff
@@ -146,7 +145,7 @@ def _handler(self, request, response):
             log_level=loglevel,
             process_step="build_output",
         )
-        response.outputs["output"].file = build_output(config)
+        response.outputs["output"].file = get_outfile(config, "hist")
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/processes/wps_convolution.py:148]: `build_output` is a function that was used specifically for `wps_convolution`, and I found Eric made the function `get_outfile`, which uses `collect_output_files` from `wps_utils`.
>What is the expected filename we are searching for?

The filename convention is described [here](https://pcic.uvic.ca/confluence/pages/viewpage.action?spaceKey=CSG&title=Meeting+Notes+2018-05-09+-+RVIC+demo)

parameters file: CASEID.rvic.prm.GRIDID.DATE.nc
convolution file: CASEID.rvic.h0a.ENDINGDATE.nc

File: osprey/processes/wps_parameters.py [hunk 3]
```diff
@@ -114,7 +155,18 @@ def _handler(self, request, response):
...
         )
-        parameters(config, np)
+
+        if os.path.isfile(unprocessed):
+            config = read_config(unprocessed)
+        else:
+            unprocessed = unprocessed.replace("'", '"')
+            config = config_hander(
+                self.workdir, parameters.__name__, unprocessed, self.config_template
+            )
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/processes/wps_parameters.py:165]: It is necessary for the [json.load](https://github.com/pacificclimate/osprey/blob/i23-params-config-builder/osprey/utils.py#L45)

File: osprey/processes/wps_parameters.py [hunk 4]
```diff
@@ -124,7 +176,7 @@ def _handler(self, request, response):
             log_level=loglevel,
             process_step="build_output",
         )
-        response.outputs["output"].file = self.get_outfile(config)
+        response.outputs["output"].file = get_outfile(config, "params")
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/processes/wps_parameters.py:179]: https://pcic.uvic.ca/confluence/pages/viewpage.action?spaceKey=CSG&title=Meeting+Notes+2018-05-09+-+RVIC+demo#MeetingNotes2018-05-09-RVICdemo-4.RoutingParametersFile
parameters file: CASEID.rvic.prm.GRIDID.DATE.nc

File: osprey/utils.py [hunk 5]
```diff
@@ -52,45 +54,66 @@ def config_hander(workdir, unprocessed, config_template):
...
 
     except KeyError as e:
         raise ProcessError("Invalid config key provided")
 
 
-def config_file_builder(workdir, config, config_template):
+def config_file_builder(workdir, config):
     """
     This function is used for RVIC1.1.0.post1 only since the version requires Configuration input
     to be a .cfg filepath. The function uses information from config to create the file.
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/utils.py:77]: I removed the function in another [PR](https://github.com/pacificclimate/osprey/pull/32). Do you wish to see it removed here as well?

File: osprey/utils.py [hunk 6]
```diff
@@ -52,45 +54,66 @@ def config_hander(workdir, unprocessed, config_template):
...
+        for upper_key in config.keys():
             cfg_file.write(f"[{upper_key}]\n")
-            for k, v in config_template[upper_key].items():
-                cfg_file.write(f"{k}: {str(v)}\n")
+            for k, v in config[upper_key].items():
+                if isinstance(v, Iterable) and type(v) != str:
+                    v_string = ", ".join(v)
+                else:
+                    v_string = str(v)
+                cfg_file.write(f"{k}: {v_string}\n")
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/utils.py:88]: >  I assume this config file format is very rigid?

Yes, it is. Reading this comment, I think I should remove the function. It is redundant and makes it harder to read.

File: osprey/utils.py [hunk 7]
```diff
@@ -52,45 +54,66 @@ def config_hander(workdir, unprocessed, config_template):
...
-    directory = os.path.join(config["OPTIONS"]["CASE_DIR"], "hist")
-    filename = ".".join([case_id, "rvic", "h0a", end_date, "nc"])
-    return os.path.join(directory, filename)
-
-
-def run_rvic(rvic_module, version, config, config_file):
+def run_rvic(workdir, rvic_module, version, config):
     if version == "1.1.0-1":  # RVIC1.1.0.post1
-        cfg_filepath = config_file
+        cfg_filepath = config_file_builder(workdir, config)
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/utils.py:95]: Not anymore, I will remove it right away.

File: osprey/utils.py [hunk 8]
```diff
@@ -52,45 +54,66 @@ def config_hander(workdir, unprocessed, config_template):
...
+
+def get_outfile(config, dir_name):
+    case_id = config["OPTIONS"]["CASEID"]
+    case_dir = config["OPTIONS"]["CASE_DIR"]
+    if dir_name == "params":
+        grid_id = config["OPTIONS"]["GRIDID"]
+        date = datetime.now().strftime("%Y%m%d")
+        filename = ".".join([case_id, "rvic", "prm", grid_id, date, "nc"])
+
+    elif dir_name == "hist":
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/utils.py:109]: It is automatically created by `rvic` using the config inputs provided.

**Comment by sum1lim** [comment]
>With the rvic version update it would be helpful to know what we need to do still (modification-wise) to get the bird working properly. Is it the same as before? Or has it changed in some way?

@nikola-rados For this PR, we still need to do both modifications in [README](https://github.com/pacificclimate/osprey/tree/i23-params-config-builder#testing)

**Comment by nikola-rados** [comment]
So the upgrade to `1.1.1` did not solve the installation problems at all? That's a shame.

File: osprey/utils.py [hunk 9]
```diff
@@ -52,45 +54,49 @@ def config_hander(workdir, unprocessed, config_template):
...
 
 
-def config_file_builder(workdir, config, config_template):
+def get_outfile(config, dir_name):
     """
-    This function is used for RVIC1.1.0.post1 only since the version requires Configuration input
-    to be a .cfg filepath. The function uses information from config to create the file.
+    This function returns the output filepath of RVIC processes.
+    Parameters
+        1. config (dict)h: Set of key-value pairs that contains the information of filename
```
Comment (nikola-rados) [inline osprey/utils.py:78]: There is a typo in here, _(dict)_.

**Review by nikola-rados — APPROVED** [review]
Looks great! Merge away.

**Comment by sum1lim** [comment]
>So the upgrade to 1.1.1 did not solve the installation problems at all? That's a shame.

That is true. Both issues were captured after 1.1.1 release (Feb, 2017), and it has never been published to pypi server. The version we are using is published in [PCIC's pypi server](https://pypi.pacificclimate.org/simple/rvic/)

**Pull request merged**